server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java

```diff
@@ -1347,10 +1365,81 @@ public Builder copy() {
         return new Builder(this);
     }
 
+    public record DataStreamIndices(
+        String namePrefix,
+        List<Index> indices,
+        boolean rolloverOnWrite,
+        @Nullable DataStreamAutoShardingEvent autoShardingEvent
+    ) {
+
+        public Index getWriteIndex() {
+            return indices.get(indices.size() - 1);
+        }
+
+        public String provideName(String dataStreamName, long generation, long epochMillis) {
```

Comment:
Does it make sense for us to ever call this method with a dataStreamName that is different from the one that this record belongs to? That said, I don't know if it's a good idea necessarily to move the data stream name into the record since it just adds another place that would need to be updated if we ever change a data stream name. Should we just make this method private to this class? Why would you ever choose to call this instead of DataStream#getDefaultIndexName?

Reply:
I played around with this, as I wasn't 100% a fan of having to pass the `dataStreamName` either. I tried making the `DataStreamIndices` class non-static so that it would be able to access the `name` field directly, but I couldn't get that to work because we need to initialize the `DataStreamIndices` during the construction of the `DataStream` object. [This](https://github.com/nielsbauman/elasticsearch/blob/f45fd2df5b095a4b36dff94d48b84deb0f26931c/server/src/main/java/org/elasticsearch/cluster/metadata/DataStream.java#L524-L524) is currently the only usage of `provideName`, which allows us to avoid having to do conditionals for example [here](https://github.com/nielsbauman/elasticsearch/blob/f45fd2df5b095a4b36dff94d48b84deb0f26931c/server/src/main/java/org/elasticsearch/action/admin/indices/rollover/MetadataRolloverService.java#L208-L208).

Reply:
Since the `provideName` method is only ever used inside the DataStream class can we limit its visibility to private or package private?